verified_boot_app.md

```diff
@@ -57,12 +173,112 @@ Src: https://github.com/Kicksecure/kicksecure-wiki-backup/blob/6900fdfbdea8693fd
+
+<!--
+* Technical high-level overview:
+** Power on -> find list of candidate bootloaders (or let user choose bootloader to load via boot menu, as usual)
+** Loop through the list trying to verify and boot each one
+** Verify bootloader signature against in-firmware trusted Secure Boot database
+** If verification passes, continue boot
+** EDIT: If verification fails, pass control to the application
+-->
+
+3.4. The firmware MAY run the application if the bootloader image verification
+     fails and the firmware is unable to boot the operating system.
+
+<!--
+There will be no enforcement of image verification if permissive policies or
+bugs like MSI inSecure Boot will be present in the firmware.
+-->
+
+3.5. The firmware SHOULD NOT implement insecure image verification policies
+     allowing to bypass signature verification when UEFI Secure Boot is
+     enabled.
+
+## Usecase examples
+
+<!--
 * Technical high-level examples:
 ** Power on -> firmware detects Microsoft Windows (bootloader) is installed -> show the key to the user -> user confirms or denies the key -> continue boot (if confirmed, otherwise go back to firmware menu or something)
 ** Power on -> firmware detects a "standard" Linux distribution such as Debian without a key in the well-known location -> show default Microsoft signing key for shim to the user -> user confirms or denies the key -> continue boot (if confirmed, otherwise go back to firmware menu or something)
 ** Power on -> firmware detects an "augmented" Linux distribution such as Kicksecure with a key database in the well-known location -> show key from key database to the user -> user confirms or denies the key -> continue boot (if confirmed, otherwise go back to firmware menu or something)
 ** User disables Secure Boot -> as usual
-* Additional notes:
-** To handle the situation where multiple shims are installed, the firmware may need the ability to find multiple candidate bootloaders in the removable media path (or any bootloader path)
+-->
 
 -->
+
+### Initial configuration
+
+<!-- TODO: diagrams? -->
+
+1. User powers on the platform.
+2. Firmware performs boot media enumeration.
+3. Firmware detects initial boot or boot with default settings and launches an
+   application.
+4. The application visits all handles with `gEfiSimpleFileSystemProtocolGuid`
+   and `gEfiPartTypeSystemPartGuid`.
+5. For each handle, find operating system bootloader. E.g. GRUB, SHIM, Windows
+   Boot Manager.
+6. For each handle, search for files containing keys and certificates. E.g.
+   files with `.der`, `.cer`, `.key` extensions.
+7. For each bootloader found in step 5, match a key found in step 6 or
+   `PKDefault`, `KEKDefault`, `dbDefault`, `dbxDefault`.
+8. If key not in `dbxDefault`, attempt image verification and report the
+   result to the user.
+9. If verification passed, ask the user to trust the key and optionally offer
+   to add a boot option.
+11. If the user decides to trust the key, the key is added to `DB`.
```

Comment:
What if the user chooses not to trust?

Reply:
Loop to the next bootloader.

Reply:
It should be reflected in documented flow.